### DISCUSSION TITLE
jwt vs session-id

### DISCUSSION THREAD

**Discussion by s103071049** [Lidemy/forum#238] [其他問題] [answered]
最近在準備面試問題，查閱了 jwt 相關資料，但不確定自己的理解是否正確。想請大家幫忙看看我對這些觀念的理解是不是有誤 ? 

jwt 是由三個部分組成，header (決定是甚麼類型的演算法與甚麼型態實做出 jwt) + payload (有哪些資料)  + signature (驗證正確性)

因為 payload 與 header 是透過 base64 進行加密，所以就可以解密得到 payload 的資訊。因此我們不會在 payload 放一些敏感資訊，比方密碼。

所以 jwt 要怎麼使用呢 ? 以下通通都以登入機制做說明

第一次登入將內容放在 body 帶到 server，server 依照帶的東西和 db 做驗證，db 確定資料正確後 server 會產生一個 jwt 給 token 給 client。只要時間沒有到期，client 將 token 帶給 server，server 就可以使用 token 與 db 做驗證，具體的說會用 authorization: bearer token 帶。

接著就是我比較不太理解的部分：jwt 的原理。

jwt 由三個東西組成：header + payload + signature(透過簽章了解東西是否有被竄改)
signature 是透過 header + payload + secret 進行 hash。header 決定我們用哪個演算法、有哪些資料型態，payload 決定我們有那些資料。如果有一天有人拿到 token 他想竄改，他發現做不到，因為簽名不對。

細看：header.payload.signature (透過 . 三個部分拼成字串) ===  jwt_token，所以現在改動 token，但因為原本 header 沒動、payload 沒動，server 端的 secret 也沒動，所以 signature 也不會變，這就導致 server 端核發的 token 與你竄改的 token 不同。你就沒辦法使用你竄改的 token 登入。(Q1: 這邊的理解正確嗎 ? )

查閱很多資料都表示，jwt 可以透過無狀態驗證。

無狀態驗證是說不需要 database 就可以驗證 token 是否正確，db 不需要儲存狀態。細看：jwt_token === header字串.payload字串.signature字串，又因為 header + payload + secret 經過 hash 會是 signature 所以與 jwt_token 的 signature 字串比對，就可以進行狀態驗證。這樣就不用在資料庫存帳密。(Q2: 這邊的理解正確嗎 ? )

這種認證不認人的方式，跟我們的好朋友 session id 很像吧 ? 

讓我們來超級比一比。

可以透過 session id 與 session data 進行關聯，session id 可以存在檔案、cookie、database，他基本上會是一串亂數。然後又是我不太確定的部分了。 

如果將 session id 存在 cookie，我的具體動作會是：我將 username、password 從 body 帶到 server，如果 db 的 users table 有這兩個 username、password，那麼我就會透過 server 產生一組亂碼 seesion_id 透過 set-cookie 存到 cookie 去，同時將這組 tooken 也存到 users 表中。因為每次登入都要將 seesion_id 與資料庫的 token 做比對，所以 session_id 不算無狀態驗證 ? (Q3: 這邊的理解正確嗎 ?)

Q4 所以我可以說 jwt 的其中一項優點是不用反覆的打 db ? 因為她是無狀態驗證 ? 

Q5 因為認證不認人，所以窩只要 server 發證出去，不管中間他更改帳密或是甚麼，在 token 沒過期的前提下，就算是舊的 jwt_token、session_id 也是會有效的進行登入 ?

Q6 當流量超大時，因為 session_id 必須要和 session data 關聯，但我又不能確定他每次進來的 server 都相同，所以就會變得超級無敵麻煩。解決方式：一、partition，高雄人就用 server1、台南人就用 server2，二、透過快取機制，紀錄這些 session_id  三、透過 load-balancer 記憶。但使用 jwt 就會變得很簡單 ? 獨立一台 server 做登入驗證，只要帶上 token 想做神麼就可以做甚麼了 ?

Q7 因為 session_id 存在 db，所以我可以輕易地去刪掉她。假設被駭客攻破，我就把資料庫的 session_id 都通通清空，但是對 jwt 而言，他是無狀態驗證，所以做不到這件事情。你發了就是發了，只能乖乖等到他過期 ? 不管中間發生甚麼事情 ? 所以站在資安的角度，session_id 優於 jwt ?

Q8 參考資料二有說，對大型分散系統來說( large-scale distributed system)，金鑰的保存很不容易。這邊我不是很理解。如果可以獨立一台 server 做登入驗證，secret 就放在他身上。我少考慮了甚麼狀態 ? 好像怪怪的

謝謝大家的幫忙 !

參考資料：
一、https://www.ruanyifeng.com/blog/2018/07/json_web_token-tutorial.html
二、https://dev.to/getd/json-web-tokens-jwt-vs-sessionid-explained-in-2-mins-1mcn

**Comment by aszx87410** [comment, marked as answer]
> 因為 payload 與 header 是透過 base64 進行加密，所以就可以解密得到 payload 的資訊。因此我們不會在 payload 放一些敏感資訊，比方密碼。

base64 不是加密，是編碼，兩個是完全不一樣的唷

> 細看：header.payload.signature (透過 . 三個部分拼成字串) === jwt_token，所以現在改動 token，但因為原本 header 沒動、payload 沒動，server 端的 secret 也沒動，所以 signature 也不會變，這就導致 server 端核發的 token 與你竄改的 token 不同。你就沒辦法使用你竄改的 token 登入。(Q1: 這邊的理解正確嗎 ? )

你的理解怪怪的，你既然都改動 token 了，要嘛是改動 header，不然就 payload 或是 signature，不可能三個都沒動
你如果改 header/payload，signature 驗證就會失敗，改了 signature 本身的話驗證也會失敗
所以除非你知道 jwt 的密鑰是什麼，不然你無法改出一個可以通過驗證的 jwt token

> 無狀態驗證是說不需要 database 就可以驗證 token 是否正確，db 不需要儲存狀態。細看：jwt_token === header字串.payload字串.signature字串，又因為 header + payload + secret 經過 hash 會是 signature 所以與 jwt_token 的 signature 字串比對，就可以進行狀態驗證。這樣就不用在資料庫存帳密。(Q2: 這邊的理解正確嗎 ? )

不對，資料庫一定是要存帳密的，否則登入功能該怎麼實作？你好像把 JWT 理解成是帳密的替代方式了，不是這樣，JWT 可以拿來跟 session 比較，跟帳密無關。不需要 DB 指的是原本如果是 session 機制的話，server 收到 session id 以後要去資料庫或是檔案比對這個 session id 對應到的資料，但是 JWT 不用，直接把 JWT 解開來就好

> 因為每次登入都要將 seesion_id 與資料庫的 token 做比對，所以 session_id 不算無狀態驗證 ? (Q3: 這邊的理解正確嗎 ?)

正確

> Q4 所以我可以說 jwt 的其中一項優點是不用反覆的打 db ? 因為她是無狀態驗證 ?

在驗證 jwt 是否合法的這一段不需要打 DB

> Q5 因為認證不認人，所以窩只要 server 發證出去，不管中間他更改帳密或是甚麼，在 token 沒過期的前提下，就算是舊的 jwt_token、session_id 也是會有效的進行登入 ?

基本上是這樣，除非你有做其他機制去做檢查

> Q6 當流量超大時，因為 session_id 必須要和 session data 關聯，但我又不能確定他每次進來的 server 都相同，所以就會變得超級無敵麻煩。解決方式：一、partition，高雄人就用 server1、台南人就用 server2，二、透過快取機制，紀錄這些 session_id 三、透過 load-balancer 記憶。但使用 jwt 就會變得很簡單 ? 獨立一台 server 做登入驗證，只要帶上 token 想做神麼就可以做甚麼了 ?

也不用獨立一台 server 做登入驗證，你 server1, server2 都可以用 jwt 做驗證

> Q7 因為 session_id 存在 db，所以我可以輕易地去刪掉她。假設被駭客攻破，我就把資料庫的 session_id 都通通清空，但是對 jwt 而言，他是無狀態驗證，所以做不到這件事情。你發了就是發了，只能乖乖等到他過期 ? 不管中間發生甚麼事情 ? 所以站在資安的角度，session_id 優於 jwt ?

差不多是這樣，所以有些人會自己實作 jwt 的手動過期機制，詳情可以搜尋一下

> Q8 參考資料二有說，對大型分散系統來說( large-scale distributed system)，金鑰的保存很不容易。這邊我不是很理解。如果可以獨立一台 server 做登入驗證，secret 就放在他身上。我少考慮了甚麼狀態 ? 好像怪怪的

獨立一台 server 做登入驗證，那如果那台 server 掛掉怎麼辦？登入就壞了，在分散式系統來說這種叫做 single point of failure，是要避免的東西，所以如果你要驗證的話至少要有兩台 server，而且後面如果只接到一個 DB，那也會有同樣問題，所以至少要有兩個 DB 互相備援，如此一來你就需要至少兩台 server + 兩個 DB，然後還有 DB 同步的問題要考慮

**Reply by s103071049** [reply]
問題二問得不太好，這個問題背後想問的是 server 是如何知道 jwt 的這個 token 是錯誤的 ? 

如果我竄改的這個 token，server 要怎麼知道這個 token 不是他核發的 token ? 

server 是每次都會將 token 的 header 與 payload 和自己保存的 secret 去做 hash 計算出一份 server 的簽章，然後再去和帶進來 token 所帶的簽章做比較，如果兩個簽章是一樣的，就代表這個 token 是正確的。如果不一樣就代表這個 token 是被偽造的 ?

jwt 是透過這樣的機制去知道這個 token 的正確與否嗎 ? 

謝謝老師的回答，其他部分有變得比較清楚!!

**Reply by aszx87410** [reply]
是，就是這樣的機制沒錯

**Reply by s103071049** [reply]
老師，謝謝你!!! 超級感溫
